### ISSUE TITLE
[flutter_webview] Cache "initialURL" for offlineview or PWA support #561

### ISSUE THREAD

**Issue by rebootcode** [flutter/flutter#69502]
## Use case

Is it possible to cache "`initialURL`" so that if there is no internet, InitialURL always opens the page from the local cache if the app has been opened at least once and the cache is not cleared.

## Proposal

Caching the `intialURL` as well as follow up URL after the first-page load just like PWA insides flutter.

**Comment by iapicca** [comment]
@rebootcode 
maybe I didn't understand correctly, but
it seems to me that you want to cache the whole destination of the url rather than the url itself
if so, I'm not sure it's realistic expectation

**Comment by rebootcode** [comment]
PWA has an offline store option and if `flutter_webview` can support PWA, It will be just a magical moment.

**Comment by sidrao2006** [comment]
@rebootcode ,To use `flutter_webview`, you would have to create a flutter app which supports android and iOS only.
`flutter_webview` currently doesn't support flutter web.
[This Stackoverflow question](https://stackoverflow.com/questions/58150503/webview-in-flutter-web) has a workaround.
After following that, building a flutter web app will enable caching by default and if you deploy the web app with SSL, it can be used as a PWA.

**Comment by rebootcode** [comment]
@sidrao2006 - Do  you mean `https://pub.dev/packages/flutter_webview_plugin` has `PWA` support using below options:

```
    bool withLocalStorage: true,
    bool withJavascript: true,
    bool clearCache: false,
    bool clearCookies: false,
    bool enableAppScheme: true,
    bool withLocalStorage: true,
    bool appCacheEnabled: false,
```

Does `enableAppScheme` implies PWA support?

**Comment by sidrao2006** [comment]
@rebootcode , sorry I'm not sure about that package

**Comment by rebootcode** [comment]
> @rebootcode , sorry I'm not sure about that package

Where is a workaround in the question, it just tells you how to create "webview" for which we already have a plugin?

**This whole thread is about supporting the "PWA" inside flutter_webview package.**

**Comment by sidrao2006** [comment]
@rebootcode , as I've mentioned above, `flutter_webview` does not support web. It throws a PlatformException if you try to run it in the web.
The question I linked to has a _workaround_ to this problem. The problem being that the authors of the package did not add web support to it and you can only run it on other platforms as a flutter app and not a PWA. Since, your aim is building a PWA, you will might to follow the given solution.
Thanks

**Comment by sidrao2006** [comment]
Also, since this is neither a bug nor a feature request, please close this issue if you agree with me
Thanks

**Comment by rebootcode** [comment]
@sidrao2006 - what do you mean, this is not a feature request?

It is a feature request for "PWA Support" inside flutter_webview plugin

**Comment by sidrao2006** [comment]
@rebootcode please check #45295 which requests to add web support for webview

**Comment by rebootcode** [comment]
> @rebootcode please check #45295 which requests to add web support for webview

I literally have no IDEA, why would people want "WebView"for flutter web. This is out of my imagination right now to imagine a `web` application inside WebView. It sounds like someone just need to learn `iframe`, but that is beyond the discussion for this thread.

**Comment by sidrao2006** [comment]
Flutter web completely supports PWAs. Flutter webview is like any other widget which can be used in a flutter app. However it isn't supported on the web (which is addressed in the above mentioned issue)

**Comment by sidrao2006** [comment]
@rebootcode Flutter already supports PWAs.

Also, one of the ways of using webview in web is with iFrame.

**Comment by rebootcode** [comment]
> @rebootcode Flutter already supports PWAs.

No, it is not, I just created a simple WebView flutter app and accessed web page, closed app, closed internet, open app and voila got the error, the page is not available.

**Comment by sidrao2006** [comment]
@rebootcode , I think you've got it a little wrong.
PWA is a standalone web app / web site with a platform wrapper provided  by the browser. You **DO NOT** have to build a native app. That's the whole point of a PWA! You just way to work with web technologies and the browser does the work for you.
What you're referring to is something like a TWA which can be deployed to the playstore as an android app but is actually a native webview.
I'm not sure whether you want something like this.
Please refer to [this Stackoverflow question](https://stackoverflow.com/a/59205480/11626847). It should satisfy your requirement.
Flutter Web supports PWA but Webviews can't support PWA. Because you cannot have a PWA inside a native app. Also, webviews support for web has to use Iframes. You cannot run a browser instance within a browser

**Comment by rebootcode** [comment]
@sidrao2006 - buddy, no offence, but you are talking all stuff that's just irritating, provoking and written with less thought or maybe someone put you on disk to write it here even though if you have no idea,

1). If you don't need it, don't come to this thread and share your input, it makes the whole thread a trash post
2). **If You think you don't need PWA with your logic you don't even need "webView"**, you can just ask your user to do "Bookmarking" of your web on the browser.

The title clearly mentions 3 interesting points:

1). PWA -> Offline access -> Mobile devices

If webview supports "local storage", "cookies" and the whole world, it must also support PWA for "offline access" property which is such a necessity.

PWA Offline access supports in `webview` is pretty powerful as it will make webview more like a native app with its offline access property. 

The main thing here is "**offline access" feature of PWA**.

**Comment by sidrao2006** [comment]
@rebootcode , I have provided a link to a package that does the same.

**Comment by sidrao2006** [comment]
Again, you are using the wrong terms and that's confusing.

PWAs are standalone. 
Yes, I agree, caching webview URLs is useful but that doesn't make it a PWA if you are using native code to implement other functionality as well.
Again, PWAs are generated by the browser, not the developer

**Comment by sidrao2006** [comment]
Please refer to this guide - https://web.dev/progressive-web-apps and this one - https://developer.mozilla.org/en-US/docs/Web/Progressive_web_apps

**Comment by Sammius** [comment]
Due to Flutter's limitations on 3D, I also have a significant portion of the application in WebView, and PWA support for WebView would be very helpful. Since in our region, there are problems with the Internet coverage.

Offtopic: dear @sidrao2006 your opinion is understandable, but it seems to me that your comments (more than half of all others) somewhat clog this thread. Sorry

**Comment by rebootcode** [comment]
@sidrao2006 - Please keep your comment off this thread, your comments are way too much distracting to the post with your little knowledge and opinion.
Kindly let the flutter team have a look at it. 
**Mr. @sidrao2006 , I highly recommend deleting your post, it's confusing to the community here**, we all are here well knowledgable developers at least know, "how to google & search", thus no point in sharing a technology glossary list.

**Comment by darshankawar** [comment]
@rebootcode 
I see there's an [open](https://github.com/flutter/flutter/issues/30582) feature request to integrate PWA for `webview_flutter` plugin. Also see [this](https://github.com/flutter/flutter/issues/30582#issuecomment-480517336) comment on feasibility of this request implementation. Would you be okay to follow-up in that issue and close this one as duplicate ?
Thanks.

**Comment by rebootcode** [comment]
@darshankawar, I see the previous [#30582](https://github.com/flutter/flutter/issues/30582) thread recommends downloading the PWA app inside flutter. 
But I recommend instead of downloading another app, just supporting the "offline" or "page cache" features using `serviceworker` or something similar inside webview to let page accessible offline.

It sounds similar and the objective may sound similar, but the approach and requirement are strictly for "offline" based access using `serviceworker` cache or something similar. 
Not sure, if merging both and closing this thread would be right.

**Comment by darshankawar** [comment]
Related to [30582](https://github.com/flutter/flutter/issues/30582)

**Comment by StarMonk** [comment]
+1 

@FlutterIssues This should be hot for webview

**Comment by bahung1221** [comment]
+1

It's necessary for Web Apps to create an experience like the "native app".

**Comment by ubaidakmal** [comment]
is there any method for offline cache web view support for the Flutter ios app?

**Comment by kyvaith** [comment]
The idea is to have something like https://www.pwabuilder.com/ but in Flutter. I'd love to see this feature here.
